contracts/MerkleAirdrop.sol

```diff
@@ -1,51 +1,113 @@
+
 /**
- * @dev A contract allow recipients to claim tokens via merkle airdrop.
+ * @dev A contract that allows recipients to claim tokens via 'merkle airdrop'.
  */
-contract MerkleAirdrop is ERC20 {
-    bytes32 public immutable merkleRoot;
+contract MerkleAirdrop is ReentrancyGuard {
+    error AirdropInActive();
+    error NotInMerkleTree();
+    error AlreadyClaimed();
+    error NotOwner();
+
+    address public owner;
+    address public sender;
+    IERC20 public token;
+    bytes32 public merkleRoot;
 
-    mapping(address => bool) public hasClaimed; // use uint8
+    mapping(address => uint8) public hasClaimed;
 
-    event Claimed(address indexed to, uint amount);
+    event Claimed(address indexed claimant, uint amount);
 
     /**
      * @dev Constructor
+     */
+    constructor() {
+        owner = msg.sender;
+    }
+
+    /**
+     * @dev initialzer
+     * @param _sender The account from which the tokens are airdropped.
+     * @param _token The address of the airdrop token.
      * @param _merkleRoot The root of the merkle tree contraining the selected recipients
      */
-    constructor(bytes32 _merkleRoot) ERC20("Merkle Airdrop Token", "MAT") {
-        merkleRoot = _merkleRoot;
+    function initialize(
+        address _sender,
+        address _token,
+        bytes32 _merkleRoot
+    ) external {
+        onlyOwner();
+        setSender(_sender);
+        setToken(_token);
+        setMerkleRoot(_merkleRoot);
     }
 
     /**
      * @dev Claims airdropped tokens.
-     * @param to The account being claimed for.
      * @param amount The amount of the claim being made.
      * @param proof A merkle proof proving the claim is valid.
      */
-    // to not required
-    // msg.sender
-    function claim(address to, uint amount, bytes32[] calldata proof) external {
-        require(!hasClaimed[to], "Already claimed");
+    function claim(
+        uint amount,
+        bytes32[] calldata proof
+    ) external nonReentrant {
+        if (
+            sender == address(0) ||
+            address(token) == address(0) ||
+            merkleRoot == bytes32(0)
+        ) {
+            revert AirdropInActive();
+        }
+        if (hasClaimed[msg.sender] != 0) {
+            revert AlreadyClaimed();
+        }
 
         // Verify merkle proof, or revert if not in tree
-        bytes32 leaf = keccak256(abi.encodePacked(to, amount));
+        bytes32 leaf = keccak256(abi.encodePacked(msg.sender, amount));
         bool isValidLeaf = MerkleProof.verify(proof, merkleRoot, leaf);
-        require(isValidLeaf, "Not in merkle tree");
+
+        if (!isValidLeaf) {
+            revert NotInMerkleTree();
+        }
 
         // Set address to claimed
-        hasClaimed[to] = true;
+        hasClaimed[msg.sender] = 1;
 
-        // Mint tokens to address
-        _mint(to, amount);
+        // Transfer tokens to msg.sender address
+        SafeERC20.safeTransferFrom(token, sender, msg.sender, amount);
 
```

Comment:
instead of writing like this, declare at the starting of this contract as follows:
`using SafeERC20 for IERC20`
and then use the SafeERC20 wrapper as follows:
`token.safeTransferFrom(...args)`